compiler/next/include/chpl/uast/IndexableLoop.h

```diff
@@ -0,0 +1,123 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_UAST_INDEXABLELOOP_H
+#define CHPL_UAST_INDEXABLELOOP_H
+
+#include "chpl/queries/Location.h"
+#include "chpl/uast/Loop.h"
+#include "chpl/uast/VariableDecl.h"
+
+namespace chpl {
+namespace uast {
+
+
+/**
+  This abstract class represents an indexable loop.
+ */
+class IndexableLoop : public Loop {
+ protected:
+  IndexableLoop(ASTTag tag, ASTList children, int8_t iterandChildNum,
+                int8_t indexVariableDeclStartChildNum,
+                int numIndexVariables,
+                bool usesDo)
+    : Loop(tag, std::move(children),
+           // Compute the start of the loop body.
+           ((iterandChildNum >= 0) + numIndexVariables),
+           usesDo),
+      iterandChildNum_(iterandChildNum),
+      indexVariableDeclStartChildNum_(indexVariableDeclStartChildNum),
+      numIndexVariables_(numIndexVariables) {
+
+    assert(isExpressionASTList(children_));
+  }
+
+  bool indexableLoopContentsMatchInner(const IndexableLoop* other) const {
+    return this->loopContentsMatchInner((const Loop*)other);
+  }
+
+  void indexableLoopMarkUniqueStringsInner(Context* context) const {
+    this->loopMarkUniqueStringsInner(context);
+  }
+
+  int8_t iterandChildNum_;
+  int8_t indexVariableDeclStartChildNum_;
+  int numIndexVariables_;
+
+ public:
+  virtual ~IndexableLoop() override = 0; // this is an abstract base class
+
+  /**
+    Return the number of index variables introduced by this indexable loop.
+  */
+  int numIndexVariables() const {
+    return numIndexVariables_;
+  }
+
+  /**
+    Returns the i'th index variable declaration of this indexable loop.
+  */
+  const VariableDecl* indexVariableDecl(int i) const {
+    if (!numIndexVariables_) return nullptr;
+    assert(0 <= i && i < numIndexVariables_);
+    return (const VariableDecl*)child(indexVariableDeclStartChildNum_+i);
+  }
+
+  /**
+    Returns the first index variable declaration of this indexable loop.
+  */
+  const VariableDecl* indexVariableDecl() const {
+    if (!numIndexVariables_) return nullptr;
+    assert(numIndexVariables_);
+    return indexVariableDecl(0);
+  }
+
+  /**
+    Returns the i'th index variable of this indexable loop.
+  */
+  const Variable* indexVariable(int i) const {
+    if (!numIndexVariables_) return nullptr;
+    assert(0 <= i && i < numIndexVariables_);
+    return this->indexVariableDecl(i)->variable();
+  }
+
+  /**
+    Returns the first index variable of this indexable loop.
+  */
+  const Variable* indexVariable() const {
+    if (!numIndexVariables_) return nullptr;
+    assert(numIndexVariables_);
+    return indexVariable(0);
+  }
+
+  /**
+    Returns the iterand of this indexable loop.
+  */
+  const Expression* iterand() const {
+    if (iterandChildNum_ < 0) return nullptr;
+    return (const Expression*)child(iterandChildNum_);
+  }
```

Comment:
Again, OK for now, however we will need to support zippering as well. Given that zippering does not nest, it makes sense to have each zippered iterand be a direct child of Indexable. (I am contrasting index variables here.)

Note that (loop is zippered) is not necessarily the same as (#iterables > 1), so need a dedicated boolean for that.

Reply:
Can't you just figure out if the loop is zippered by looking at the iterand?

Reply:
I agree with @vasslitvinov here that we should have an interface here to get the i'th zippered iterand etc. Zippering is syntactical so should be reflected here.

Of course we could do something like introduce a `Zip` uAST node but since it only comes up in loops and there can only be one per loop I think it makes more sense to have `iterand(int i)` and `numIterands()` etc.

Reply:
This whole time I assumed that there would be a `Zip` AST node. I think it appears in often enough places that we could just make a `Zip` as a subclass of `Call`.